src/avro/src/decode.rs

```diff
@@ -569,66 +580,195 @@ pub mod public_decoders {
             impl AvroDecode for $name {
                 type Out = $out;
                 fn scalar(self, scalar: Scalar) -> Result<$out, Error> {
+                    use std::convert::TryInto;
                     let out = match scalar {
                         $(
-                            Scalar::$scalar_branch(inner) => {inner as $out}
+                            Scalar::$scalar_branch(inner) => {inner.try_into()?}
                         ),*
                             other => bail!("Unexpected scalar: {:#?}", other)
                     };
                     Ok(out)
                 }
             }
+
+            impl AvroDecodeable for $out {
+                type Decoder = $name;
+                fn new_decoder() -> $name {
+                    $name
+                }
+            }
         }
     }
 
     define_simple_decoder!(I32Decoder, i32, Int;Long);
-    define_simple_decoder!(I64Decoder, i64, Long);
+    define_simple_decoder!(I64Decoder, i64, Int;Long);
+    define_simple_decoder!(U64Decoder, u64, Int;Long);
+    define_simple_decoder!(UsizeDecoder, usize, Int;Long);
```

Comment:
I think that this is the same thing that you refer to inside the proc macro as trivial_decoder, if so it might be nicer to re-name that to simple_decoder.

Reply:
It's not, actually. Trivial decoders return `()`, whereas these return `i32`, `i64`, etc.

The reason trivial decoder exists is because you still need to decode a record, even if you're discarding the result, to know how many bytes to skip in e.g. an OCF file (because Avro isn't length-delimited...)